### ISSUE TITLE
Difference Ncanonical and Ndiff

### ISSUE THREAD

**Issue by SophieValk** [nanoporetech/modkit#114]
Hi, 

I have been struggling with understanding the exact difference between Ncanonical and Ndiff. 
I have done 5mC modified basecalling using Dorado on Nanopore reads and have subsequently called CpG modification using Modkit. 

When looking at the (using pileup) generated .bed file and comparing this to the sorted.bam file in IGV I have a hard time understanding the difference between Ncanonical and Ndiff. 

2 example lines of the modkit pileup .bed file look like the following:
NC_079164.1	74448320	74448321	m	6	.	74448320	74448321	255,0,0	6 100.00 6 0 0 0 2 1 0
NC_079164.1	74448322	74448323	m	6	.	74448322	74448323	255,0,0	6 83.33 5 1 0 0 0 3 0

The sorted.bam file alignment looks like this (with red blocks indicating that the cytosine is methylated):
![Screenshot from 2024-01-18 09-59-10](https://github.com/nanoporetech/modkit/assets/106058726/75838a8b-f593-4ef5-ac6a-50f3359c690e)

For example: The CpG site on location NC_079164.1:74448320-74448321 should have Nmod=6. Nfail=2, and Ndiff=1. From the same position in the sorted.bam file I count 8 methylated Cs and only one unmethylated C. From the Nfail=2 I can understand that from the 8 methylated Cs, two calls may be below the quality threshold (which I set at 0.8) and so from the 8 methylated cytosines, 2 are being ignored. And the one read that seems to have no methylated cytosine on that CpG is called as Ndiff=1. That all makes sense! 

Then the next position: The CpG site on location NC_079164.1:74448323-74448322 should have Nmod=5, Ncanonical=1, and Nfail=3, according to the modkit pileup .bed file. The sorted.bam file in IGV shows that 5 cytosines are methylated, and there are 4 reads that have a C>T substitution. So probably the Nfail=3 are failed calls for these substitutions, which makes 3 out of 4 being ignored, and so 1 Ncanonical remains. That makes sense as well. What I do not understand is that these substitutions are marked as Ncanonical, and not as Ndiff. 

However, In the first example, the modification frequency is 100 (Nmod=6 and Ncanonical=0). In the second example the modification frequency is 83.33 (Nmod=5 and Ncanonical=1). What I do not understand is that when a cytosine is unmethylated it is called as Ndiff, whereas if a base is substituted by another canonical base it is called as Ncanonical. In this way the modification frequency is influenced by the amount of (e.g. C>T) substitutions, whereas the number of unmethylated cytosines (Ndiff) does not have an influence on the methylation frequency. Do I understand this correctly? And could anyone perhaps explain to me why Ndiff is not incorporated in the methylation frequency? 

Thanks, 
Sophie

**Comment by ArtRand** [comment]
Hello @SophieValk,

For the first position:
NC_079164.1 74448320 74448321 m 6 . 74448320 74448321 255,0,0 6 100.00 6 0 0 0 2 1 0

I'll annotate the important values:
valid_cov: 6 
pct_mod: 100.00 
n_mod: 6 
n_canonical: 0 
n_other: 0 
n_delete: 0 
n_fail: 2 
n_diff: 1 
n_no_call: 0

From what I can see, you have 8 reads predicting 5mC with reasonable confidence, the 8 red blocks. I can't tell what the first (+)-strand read predicts, but sometimes IGV will make the 5mC color _very_ faint when the probability is low. There is one read that doesn't make sense with the information in the IGV shot. Either that top read has a 5mC prediction that failed or it has a mismatch that's not showing up for some reason. 

The second position actually makes more sense:
NC_079164.1 74448322 74448323 m 6 . 74448322 74448323 255,0,0 6 83.33 5 1 0 0 0 3 0

Again, pulling out the important values:
valid_cov: 6 
pct_mod: 83.33 
n_mod: 5 
n_canonical: 1 
n_other: 0 
n_delete: 0 
n_fail: 0 
n_diff: 3 
n_no_call: 0

There 6 reads with passing 5mC calls, the 3 (+)-strand ones with the red blocks, the bottom-most (-)-strand reads and the top (-)-strand read has a canonical call. There are 3 (+)-strand reads with C>T transitions, these are the 3 n_diff. I think you mis-parsed the row and through these were n_fail. There is some pandas code on the [epi2me blog](https://labs.epi2me.io/how-to-mod/) that makes it a lot easier to read these data (I have to do this little routine every time myself). 

A couple things to try. 

1. Run `modkit extract` on this small region and see what the calls are [docs are here](https://nanoporetech.github.io/modkit/intro_extract.html#tabulating-base-modification-calls-for-each-read-position), you can use `--region NC_079164.1:74448320-74448324` and it should pull out just these reads. Check the canonical calls for position `74448320`. 
2. Make sure that IGV isn't hiding anything from you (sometimes the view can subsample reads, for example)

Could you attach a subset of the modBAM to this thread: 
```bash
$ samtools view -bh <mod_bam> NC_079164.1:74448320-74448321
```

so I can double check that first position? 

To answer your original question, N_diff is the number of reads with a base that is different from the modified base reported in the bedMethyl row. So a 5mC is a Cytosine, if a read aligns a T to this position it will be counted in N_diff.

**Comment by SophieValk** [comment]
Dear @ArtRand,

Thankyou so much for this elaborate explanation, it is really helpful!

I have tried running modkit extract on the selected region, the .tsv file shows the following results: 
**For NC_079164.1:74448320:** 
a44b8a88-0e22-4b16-951f-51a876704486	3313	74448320	NC_079164.1	+	+	+	1201	1	20802	0.9980469	m	38	.	GCCGC	C	C	false
40a6f433-7848-4903-a6dc-48a51bd717ec	2081	74448320	NC_079164.1	+	+	+	69	3	8027	0.8417969	m	25	.	GCCGT	C	C	false
351258df-e28b-41ff-8653-fb4582c2eeb7    1016    74448320	NC_079164.1	+	+	+	51	0	2999	0.9980469	m	41	.	GCCGC	C	C	false
ba095181-c78d-4d20-a9db-96ecc5ec96e4	807	74448320	NC_079164.1	+	+	+	67	111	1051	0.9980469	m	41	.	GCCGT	C	C	false
85f4eca6-bb1e-4352-a235-5bfdc8f3a8f5	187	74448320	NC_079164.1	+	+	+	79	52	5430	0.9980469	m	41	.	GCCGC	C	C	false

**For NC_079164.1:74448321:** 
9f515683-7a47-45e5-9ca6-133f9263f1dd	2442	74448321	NC_079164.1	+	-	-	66	5149	9665	0.8730469	m	17	.	CACGG	C	C	false
d90f1bd4-68da-47e7-80a3-e3bddb712922	2862	74448321	NC_079164.1	+	-	-	76	44	3284	0.9980469	m	39	.	CGCGG	C	C	false
66c480be-c18a-401e-a76d-3850984578a7	1502	74448321	NC_079164.1	+	-	-	70	54	1651	0.9980469	m	40	.	CGCGG	C	C	false

From this table it seems that there is no Ndiff read present after all. 

When I click on the first (+) read on position NC_079164.1:74448320-74448321 (the read that does not seem to contain any modification), the following details pop up: 
![Screenshot_Readinformation_19-01-2024](https://github.com/nanoporetech/modkit/assets/106058726/4016bec8-3bef-4f75-9023-23d0719e396e)
The read id: **9a217e1d-5d31-4f71-a069-dcf110cdc6c8**, does not show in the modkit extract .tsv table either. 

For position **NC_079164.1:74448324:**  , that particular read does show up: 
**9a217e1d-5d31-4f71-a069-dcf110cdc6c8**	9556	74448324	NC_079164.1	+	+	+	5196	53	17149	0.080078125	m	19	.	TGCAG	C	C	false
a44b8a88-0e22-4b16-951f-51a876704486	3317	74448324	NC_079164.1	+	+	+	1201	1	20802	0	m	50	.	CGCAG	C	C	true
40a6f433-7848-4903-a6dc-48a51bd717ec	2085	74448324	NC_079164.1	+	+	+	69	3	8027	0.103515625	m	33	.	TGCAG	C	C	false
351258df-e28b-41ff-8653-fb4582c2eeb7	1020	74448324	NC_079164.1	+	+	+	51	0	2999	0	m	40	.	CGCAG	C	C	true
ba095181-c78d-4d20-a9db-96ecc5ec96e4	811	74448324	NC_079164.1	+	+	+	67	111	1051	0.080078125	m	45	.	TGCAG	C	C	false
85f4eca6-bb1e-4352-a235-5bfdc8f3a8f5	191	74448324	NC_079164.1	+	+	+	79	52	5430	0	m	50	.	CGCAG	C	C	true

I don't really understand why the read is incorporated on base 74448324 but not on base 74448321. Even if the quality of the called read would be below the threshold, you would expect it to still be incorporated in the .tsv right? 

Hereby the modBam output of the following command:
./samtools view -bh sorted.bam NC_079164.1:74448320-74448321 >  subsetmod_filetoattach.bam 
Filesender link: **https://filesender.surf.nl/?s=download&token=73818f04-aeaf-4914-9b83-4f698a1ca721
**

**Comment by ArtRand** [comment]
Hello @SophieValk,

I downloaded the modBAM you sent (thanks) and the [genome assembly](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_947568905.1/). 

When I plot on IGV, it looks different for me:
![NC_079164_1_74448303-74448342-groupby-strand](https://github.com/nanoporetech/modkit/assets/8851152/b669632f-1b11-4825-9e5a-a3c2f39cba12)

at position NC_079164.1:74448320-74448321 I can see 8 reads with 5mC probabilities and 1 read with a C>A transversion. 

I get the same answers as you do when running 
```
modkit pileup subsetmod_filetoattach.bam - --cpg --combine-strands --ref ncbi_dataset/data/GCF_947568905.1/GCF_947568905.1_xbOstEdul1.1_genomic.fna --region NC_079164.1:74448320-74448323
```
I get:
```
NC_079164.1     74448320        74448321        m       6       .       74448320        74448321        255,0,0 6 100.00 6 0 0 0 2 1 0
NC_079164.1     74448322        74448323        m       6       .       74448322        74448323        255,0,0 6 83.33 5 1 0 0 0 3 0
```
pulling out the counts:
valid_cov: 6 
pct_mod: 100.00 
n_mod: 6 
n_canonical: 0 
n_other: 0 
n_delete: 0 
n_fail: 2 
n_diff: 1 
n_nocall: 0

**However**, I didn't set the threshold. The estimated threshold is `0.9042969`. I can inspect the read calls with the following command:
(aside, you don't need to specify the filter threshold - it will find the same number)
```
modkit extract subsetmod_filetoattach.bam null --region NC_079164.1:74448320-74448323 --read-calls extract_read_calls_region.tsv --filter-threshold 0.9042969
```
Then I narrow the output with `awk '$3==74448320 || $3==74448321' extract_read_calls_region.tsv | sort -k 3`

```
read_id                               forward_read_position  ref_position  chrom        mod_strand  ref_strand  ref_mod_strand  fw_soft_clipped_start  fw_soft_clipped_end  read_length  call_prob  call_code  base_qual  ref_kmer  query_kmer  canonical_base  modified_primary_base  fail   inferred  within_alignment
351258df-e28b-41ff-8653-fb4582c2eeb7  1016                   74448320      NC_079164.1  +           +           +               0                      51                   2999         0.9980469  m          41         .         GCCGC       C               C                      false  false     true
a44b8a88-0e22-4b16-951f-51a876704486  3313                   74448320      NC_079164.1  +           +           +               1201                   1                    20802        0.9980469  m          38         .         GCCGC       C               C                      false  false     true
ba095181-c78d-4d20-a9db-96ecc5ec96e4  807                    74448320      NC_079164.1  +           +           +               67                     111                  1051         0.9980469  m          41         .         GCCGT       C               C                      false  false     true
40a6f433-7848-4903-a6dc-48a51bd717ec  2081                   74448320      NC_079164.1  +           +           +               69                     3                    8027         0.8417969  m          25         .         GCCGT       C               C                      true   false     true
85f4eca6-bb1e-4352-a235-5bfdc8f3a8f5  187                    74448320      NC_079164.1  +           +           +               79                     52                   5430         0.9980469  m          41         .         GCCGC       C               C                      false  false     true
9f515683-7a47-45e5-9ca6-133f9263f1dd  2442                   74448321      NC_079164.1  +           -           -               66                     5149                 9665         0.8730469  m          17         .         CACGG       C               C                      true   false     true
66c480be-c18a-401e-a76d-3850984578a7  1502                   74448321      NC_079164.1  +           -           -               70                     54                   1651         0.9980469  m          40         .         CGCGG       C               C                      false  false     true
d90f1bd4-68da-47e7-80a3-e3bddb712922  2862                   74448321      NC_079164.1  +           -           -               76                     44                   3284         0.9980469  m          39         .         CGCGG       C               C                      false  false     true
```

You can see that there are 8 reads with calls and 2 fall below the threshold value of 0.9042969. Read 9a217e1d-5d31-4f71-a069-dcf110cdc6c8 _doesn't have a call_ because it didn't call C at 74448320.

So far, the IGV, pileup, and extract are copacetic. 

When I re-run `modkit pileup` with the threshold you stated (`--filter-threshold 0.8`) I get a different answer from pileup:
command:
```
modkit pileup subsetmod_filetoattach.bam - --cpg --combine-strands --ref ncbi_dataset/data/GCF_947568905.1/GCF_947568905.1_xbOstEdul1.1_genomic.fna --region NC_079164.1:74448320-74448323 --filter-threshold 0.8
```
output:
```
NC_079164.1     74448320        74448321        m       8       .       74448320        74448321        255,0,0 8 100.00 8 0 0 0 0 1 0
NC_079164.1     74448322        74448323        m       6       .       74448322        74448323        255,0,0 6 83.33 5 1 0 0 0 3 0
```

Notice that NC_079164.1:74448320-74448321 now has 8 valid reads. This makes sense because as we saw in the extract output all of the calls have probability >0.8. 

Long story short. I think your IGV is obfuscating what is going on. I use this option when looking at mods in IGV:
![image](https://github.com/nanoporetech/modkit/assets/8851152/f8f9ea93-05b9-4634-85b3-9799b6648511)

Please confirm that this all makes sense on your end. Happy to help.

**Comment by SophieValk** [comment]
Dear @ArtRand, 

Thankyou for your explanation, that indeed all makes a lot of sense. 
The confusion was indeed caused by the visualization of IGV. I have the exact same settings with coloring alignments in IGV based on all modifications. However, in View>Color Legends>Methylation, I did not have blue included and the color scaling went from white to red on a scale of 0-100. When I change this color coding from blue (range 0-50) and red (range 50-100):
![Screenshot from 2024-01-22 12-08-38](https://github.com/nanoporetech/modkit/assets/106058726/7c3fd7a5-a622-4a17-bf29-64de1f701f96)

the Ndiff suddenly pops up on site 74448320:
![Screenshot from 2024-01-22 12-08-55](https://github.com/nanoporetech/modkit/assets/106058726/982eabcd-76a1-44ba-958d-b1a4177832d9)

I don't really get how changing the color coding suddenly visualizes mutations but at least I'm now aware that I have to pay good attention at my IGV settings. 

Thanks again for helping me understand the calling specifics of Modkit and I will definitely be using this tool in the future. 

All the best, 
Sophie

**Comment by ArtRand** [comment]
Hello @SophieValk, 

No problem, happy to help. Feel free to re-open this issue if you have additional questions. For what it's worth, we're also working on improving the outputs of modkit for visualization, so keep an ear out for that.
